---
title: 'Crab-eating macaque PBMCs: Mapping human-homolog gene symbols'
author: 'Mathew Chamberlain'
date: "`r Sys.Date()`"
output:
  html_document:
    code_folding: hide
    theme: sandstone
    toc: true
    includes:
      after_body: footer.html
---

This vignette shows how to map gene symbols for PBMCs derived from Cynomolgus monkey (crab-eating macaque) to human homologs. 

Load the essential packages: Signac, Seurat, biomaRt and tidyverse.

```{r setup packages, echo = T}
library(Signac)
library(Seurat)
library(biomaRt)
library(tidyverse)
```

After mapping the reads to the Macaca fasccicularis genome, we load the genes.

```{r load genes, echo = T}
features.tsv <- read.delim("../../../../data/2021-01-04-Monkey/count/3003/outs/raw_feature_bc_matrix/features.tsv.gz", header=FALSE, stringsAsFactors=FALSE)
head(features.tsv)
```

Next, we map the gene homologs from monkey gene symbols to human gene symbols using biomaRt.

```{r map genes}
# get human and cyno gene symbols
human.R95 <- useMart(host='jan2019.archive.ensembl.org', 
				   biomart='ENSEMBL_MART_ENSEMBL', 
				   dataset='hsapiens_gene_ensembl')
cyno.R95 <- useMart(host='jan2019.archive.ensembl.org', 
				   biomart='ENSEMBL_MART_ENSEMBL', 
				   dataset='mfascicularis_gene_ensembl')
 
# values = listeENSID: list of cynomolgus ensembl IDs to be retrieved. 
listeENSID = features.tsv$V1
orthologs <- getLDS(attributes = c("ensembl_gene_id","external_gene_name"), 
				  filters = "ensembl_gene_id", 
				  values = listeENSID, 
				  mart = cyno.R95, 
				  attributesL = c("hgnc_symbol","ensembl_gene_id"), 
				  martL = human.R95)
orthologs <- as_tibble(orthologs)
colnames(orthologs) <- c("GeneID","cynoSymbol","HumanSymbol","HumanGeneID")
  
# keep only 1:1 orthologs
one2one <- orthologs %>% 
group_by(GeneID) %>% 
summarise(n()) %>% 
filter(`n()`<=1) %>%
dplyr::select(`GeneID`) %>%
pull()
orthologs <- orthologs %>% filter(GeneID %in% one2one)

# replace empty HumanSymbol (where there isn't a gene name for a homologous gene) with NA
orthologs <- orthologs %>%
mutate(HumanSymbol=replace(HumanSymbol,HumanSymbol=="", NA))
orthologs <- orthologs %>%
mutate(cynoSymbol=replace(cynoSymbol,cynoSymbol=="", NA))

idx = match(listeENSID, orthologs$GeneID)
xx = orthologs$HumanSymbol[idx]
logik = !is.na(orthologs$HumanSymbol[idx]) # sum(logik) returns 17,365 homologous genes
xx = xx[logik]
# note: several of these genes are not unique mappers; we will aggregate them later or make them unique. To aggregate, where E is the sparse expression matrix with rownames set to xx:
# E = Matrix.utils::aggregate.Matrix(E, row.names(E))
```
After mapping gene homologs, Signac can be used to classify the cell types.

## Session information

```{r sessioninfo, message = F, echo = F}
sessionInfo()
```
