---
title: 'Signac and Seurat: Multi-modal analysis of CITE-seq PBMCs from 10X Genomics'
author: 'Mathew Chamberlain'
date: "`r Sys.Date()`"
output:
  html_document:
    theme: sandstone
    toc: true
    includes:
      after_body: footer.html
---

This vignette shows how to use Signac with Seurat. There are three parts: Seurat, Signac and then visualization. We use an example CITE-seq data set from 10X Genomics.

## Seurat 

Start with the standard pre-processing steps for a Seurat object.

```{r setupSeurat, message = F, eval = F}
library(Seurat)
```

Download data from 10X Genomics.

```{r setup, message = F, eval = F}
download.file("https://cf.10xgenomics.com/samples/cell-exp/3.0.0/pbmc_10k_protein_v3/pbmc_10k_protein_v3_filtered_feature_bc_matrix.h5", destfile = "pbmc_10k_protein_v3_filtered_feature_bc_matrix.h5")
```

Create a Seurat object, and then perform SCTransform normalization. Note:

*You can use the legacy functions here (i.e., NormalizeData, ScaleData, etc.), use SCTransform or any other normalization method (including no normalization). We did not notice a significant difference in cell type annotations with different normalization methods.
*We think that it is best practice to use SCTransform, but it is not a necessary step. Signac will work fine without it.

```{r Seurat, message = T, eval = F}
# load dataset
E = Read10X_h5(filename = "pbmc_10k_protein_v3_filtered_feature_bc_matrix.h5")
pbmc <- CreateSeuratObject(counts = E$`Gene Expression`, project = "pbmc")

# store mitochondrial percentage in object meta data
pbmc <- PercentageFeatureSet(pbmc, pattern = "^MT-", col.name = "percent.mt")

# run sctransform
pbmc <- SCTransform(pbmc,vars.to.regress = "percent.mt", verbose = FALSE)
```

Perform dimensionality reduction by PCA and UMAP embedding. Note:

*Signac actually needs these functions since it uses the nearest neighbor graph generated by Seurat.

```{r Seurat 2, message = T, eval = F}
# These are now standard steps in the Seurat workflow for visualization and clustering
pbmc <- RunPCA(pbmc, verbose = FALSE)
pbmc <- RunUMAP(pbmc, dims = 1:30, verbose = FALSE)
pbmc <- FindNeighbors(pbmc, dims = 1:30, verbose = FALSE)
```

## Signac

First, make sure you have the Signac package installed.

```{r setup2, message = F, eval = F}
devtools::install_github("mathewchamberlain/Signac")
library(Signac)
```

Generate Signac labels for the Seurat object. Note:

*Optionally, you can do parallel computing by setting num.cores > 1 in the Signac function.
*Run time is ~minutes for ~10,000 cells.

```{r Signac, message = T, eval = F}
# Run Signac
data("training_HPCA")
labels <- Signac(pbmc, R = training_HPCA)
celltypes = Generate_lbls(labels, E = pbmc)
```

## Visualizations

Now we can visualize the cell type classifications at many different levels:
Immune and nonimmune
```{r Seurat Visualization 0, message = F}
pbmc <- Seurat::AddMetaData(pbmc, metadata=celltypes$Immune, col.name = "immmune")
pbmc <- Seurat::SetIdent(pbmc, value='immmune')
DimPlot(pbmc)
```

```{r Seurat Visualization 1, message = F}
pbmc <- Seurat::AddMetaData(pbmc, metadata=celltypes$L2, col.name = "celltypes")
pbmc <- Seurat::SetIdent(pbmc, value='celltypes')
DimPlot(pbmc)
```

```{r Seurat Visualization 2, message = F}
lbls = factor(celltypes$CellTypes)
levels(lbls) <- sort(unique(lbls))
pbmc <- Seurat::AddMetaData(pbmc, metadata=lbls, col.name = "celltypes")
pbmc <- Seurat::SetIdent(pbmc, value='celltypes')
DimPlot(pbmc)
```

```{r Seurat Visualization 3, message = F}
lbls = factor(celltypes$CellStates)
levels(lbls) <- sort(unique(lbls))
pbmc <- Seurat::AddMetaData(pbmc, metadata=lbls, col.name = "celltypes")
pbmc <- Seurat::SetIdent(pbmc, value='celltypes')
DimPlot(pbmc)
```

```{r Seurat Visualization 4, message = F}
lbls = factor(celltypes$CellStates_novel)
levels(lbls) <- sort(unique(lbls))
pbmc <- Seurat::AddMetaData(pbmc, metadata=lbls, col.name = "celltypes")
pbmc <- Seurat::SetIdent(pbmc, value='celltypes')
DimPlot(pbmc)
```

Add protein expression information

```{r Seurat add protein, message = F}
pbmc[["ADT"]] <- CreateAssayObject(counts = E$`Antibody Capture`[,colnames(E$`Antibody Capture`) %in% colnames(pbmc)])
pbmc <- NormalizeData(pbmc, assay = "ADT", normalization.method = "CLR")
pbmc <- ScaleData(pbmc, assay = "ADT")
```

Visualize protein information on the RNA clustering result

```{r Seurat visualize protein, message = F}
FeaturePlot(pbmc, features = c("CD8a-TotalSeqB", "CD8A"), ncol = 2)
```

```{r Seurat visualize protein 2, message = F}
RidgePlot(pbmc, features = c("CD8a-TotalSeqB"))
```

Identify differentially expressed proteins between clusters

```{r Seurat visualize protein 3, message = F}
# Downsample the clusters to a maximum of 500 cells each (makes the heatmap easier to see for
# small clusters)
pbmc.small <- subset(pbmc, downsample = 500)

# Find protein markers for all clusters, and draw a heatmap
adt.markers <- FindAllMarkers(pbmc.small, assay = "ADT", only.pos = TRUE, verbose = F)
DoHeatmap(pbmc.small, features = unique(adt.markers$gene), assay = "ADT", angle = 90) + NoLegend()
DefaultAssay(pbmc) <- "ADT"
```

Cluster the data based on protein expression information.

```{r Seurat cluster on protein, message = F}
pbmc <- FindVariableFeatures(pbmc, selection.method = "vst")
pbmc <- ScaleData(pbmc)
pbmc <- RunPCA(pbmc, verbose = F)
pbmc <- FindNeighbors(pbmc)
pbmc <- FindClusters(pbmc, resolution = 0.8)
pbmc <- RunUMAP(pbmc, dims = 1:10)
```

Visualize Signac cell types on protein clusters
```{r Seurat protein clusters vis 1, message = F}
lbls = factor(celltypes$CellTypes)
levels(lbls) <- sort(unique(lbls))
pbmc <- Seurat::AddMetaData(pbmc, metadata=lbls, col.name = "celltypes")
pbmc <- Seurat::SetIdent(pbmc, value='celltypes')
DimPlot(pbmc)
```

```{r Seurat protein clusters vis 2, message = F}
lbls = factor(celltypes$CellStates)
levels(lbls) <- sort(unique(lbls))
pbmc <- Seurat::AddMetaData(pbmc, metadata=lbls, col.name = "celltypes")
pbmc <- Seurat::SetIdent(pbmc, value='celltypes')
DimPlot(pbmc)
```

```{r Seurat protein clusters vis 3, message = F}
FeaturePlot(pbmc, features = c("CD8a-TotalSeqB", "CD8A"))
```

Save results

```{r save results, message = F, eval = F}
saveRDS(pbmc, file = "citeseq_signac.rds")
```

## Session information

```{r sessioninfo, message = F, echo = F}
sessionInfo()
```
